---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# varAnnoPrototype

<!-- badges: start -->
<!-- badges: end -->

vcf files are complex and can be difficult to interpret. varAnnoPrototype is a prototype variant annotation tool that extract some relevant information from vcf files in addition to annotating the variants contained within the file to the location and type of genomic feature in which they are located as well as the biological consequence of the variant. This package contains a set of functions to read in a vcf file, extract the relevant information, perform the genomic annotations, and output a dataframe containing the result which can be written to file, read in external software, and shared with others.

## Installation

The development version of this package can be installed from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("rgreenste/varAnnoPrototype")
```
## Example Usage

Here is an example of how you could use this package to annotate a vcf file. Please not that at this time only vcf files for human data created with the h19 reference are supported. Additional support for other genome builds including coordinate liftover could be included in future updates.

### Load the library

After installing the package load the library with the following R code:
```{r example_load, echo = T}
library(varAnnoPrototype)
## basic example code
```

### Read in a vcf fie

A sample vcf file is contained within the extdata of this package. Extract the filepath and load the file with the following code:

```{r example_readVCF, echo = T}
# extract the path to the vcf file
vcfFile<-system.file("extdata", "Challenge_data_(1).vcf", package="varAnnoPrototype")

# import the vcf file with package function importExpandedVCF()
vcfTest<-importExpandedVCF(vcfFile, genome = "hg19")

# take a quick look at the result
vcfTest
```



### Extract rowData from ExpandedVCF file


Next extract the rowData from the the ExpandedVCF object. rowData contains the genomic coordinates of the variants along with the Reference (REF) and Alternate (ALT) alleles. A column contain the variant name in the ExAC notation style (chromosome-position-REF-ALT) is created from the information in rowData:

```{r example_rowData, echo = T}
# extract info from the ExpandedVCF object
rd<-extract_rowData(vcfTest)

# take a quick look at the result
rd
```



### Extract Metrics from INFO Field of vcf File


Next extract some metrics from the INFO field of the vcf file via the ExpandedVCF object including type of variant, depth of sequencing coverage at site of variation, number of reads supporting the variant, number of of reads supporting the reference and percentage of reads supporting the variant.

```{r example_infoMetrics, echo = T}
# extract info from the ExpandedVCF object
infoMetrics<-extract_infoMetrics(vcfTest)

# take a quick look at the result
head(infoMetrics)
```


### Annotate Variants Based on Genomic Feature Location


Next extract Location information about each variant with respect to genomic features. Variant positions are queried against a file of genomic features called `TxDb.Hsapiens.UCSC.hg19.knownGene` from the R/Bioconductor package of the same name. Information about this package can be found [here] (https://bioconductor.org/packages/release/data/annotation/html/TxDb.Hsapiens.UCSC.hg19.knownGene.html). Information is returned about the location of each variant with respect to genomic features and, if located within a gene, additionally the Gene ID number for that annotation. 

Possible genomic feature locations include `coding`, `spliceSite`, `promoter`, `fiveUTR`, `threeUTR`, `intron`, and `intergenic`. 

```{r example_variantLocation, echo = T, warning = F}
# extract the variant location with respect to genomic features
varLoc<-extract_variantLocation(rd)

# take a quick look at the result
varLoc
```


### Annotate Coding Variants by Biological Consequence


Next extract information about the biological consequence of variants within coding regions of the genome. Coding variant positions with respect to genes are determined by querying against the `TxDb.Hsapiens.UCSC.hg19.knownGene` genomic feature file. The surrounding genomic sequence and nucleotide consequence of the variant is determined by querying the hg19 human genome reference sequence found the BioStrings object `BSgenome.Hsapiens.UCSC.hg19` from the R/Bioconductor package of the same name. Information about this package can be found [here] (https://bioconductor.org/packages/release/data/annotation/html/BSgenome.Hsapiens.UCSC.hg19.html). Information is returned about the consequence of each variant in the CONSEQUENCE column. Possible consequences include `frameshift`, `nonsense`, `nonsynonymous`, and `synonymous`.  

Additional information of interest about each variant is also returned, including the reference allele codon and amino acid (REFCODON, REFAA), the alternative allele codon and amino acid (VARCODON, VARAA) and the variant name in the ExAC style. 


```{r example_variantConsequence, echo = T, warning = F}
# extract the variant consequence with respect to genomic features
varConseq<-extract_variantConsequence(rd)

# take a quick look at the result
varConseq
```





